File: photonflow-cpp/src/blocks/optical/pm.cpp
/**
 * @file pm.cpp
 * @brief Phase Modulator block implementation.
 */

#include "photonflow/blocks/optical/pm.hpp"
#include "photonflow/blocks/block_registry.hpp"

#include <cmath>
#include <complex>

namespace photonflow {

PM::PM(const std::string &id, const json &params, const json &nonideal)
    : BaseBlock(id, params, nonideal) {}

std::unordered_map<std::string, Signal>
PM::process(const std::unordered_map<std::string, Signal> &inputs,
            SimContext &ctx) {
  const Signal &opt_in = inputs.at("opt_in");

  // Get electrical input
  Eigen::VectorXd v;
  if (inputs.contains("elec_in")) {
    const Signal &elec_in = inputs.at("elec_in");
    v.resize(elec_in.n_samples());
    for (Eigen::Index i = 0; i < elec_in.n_samples(); ++i) {
      v[i] = elec_in.data[i].real();
    }
  } else {
    v = Eigen::VectorXd::Zero(opt_in.n_samples());
  }

  double vpi = get_param("Vpi", 3.5);
  double phi_bias = get_param("phi_bias", 0.0);

  bool nonideal_enable = get_nonideal("enable", false);
  double bias_error = 0.0;
  double loss_db = 0.0;

  if (nonideal_enable) {
    double vpi_error = get_nonideal("vpi_error_pct", 0.0);
    vpi *= (1.0 + vpi_error / 100.0);

    bias_error = get_nonideal("bias_error_rad", 0.0);
    loss_db = get_nonideal("loss_db", 0.0);

    double drive_noise_rms = get_nonideal("drive_noise_rms", 0.0);
    if (drive_noise_rms > 0.0) {
      v += drive_noise_rms * ctx.randn(static_cast<int>(v.size()));
    }
  }

  const Eigen::Index n = opt_in.n_samples();
  Eigen::VectorXcd data(n);

  for (Eigen::Index i = 0; i < n; ++i) {
    double phi = (phi_bias + bias_error) + M_PI * v[i] / vpi;
    std::complex<double> phase_term = std::exp(std::complex<double>(0.0, phi));
    data[i] = opt_in.data[i] * phase_term;
  }

  if (nonideal_enable && loss_db != 0.0) {
    double loss_linear = std::pow(10.0, -loss_db / 20.0);
    data *= loss_linear;
  }

  Signal output(data, opt_in.fs, opt_in.t0);
  output.center_freq = opt_in.center_freq;
  output.pol_mode = opt_in.pol_mode;
  output.meta = opt_in.meta;

  return {{"opt_out", std::move(output)}};
}

std::optional<std::string> PM::port_type(const std::string &port) const {
  if (port == "opt_in" || port == "opt_out")
    return "optical";
  if (port == "elec_in")
    return "electrical";
  return std::nullopt;
}

std::optional<double> PM::estimate_fmax() const {
  if (params_.contains("bandwidth_hz")) {
    double bw = params_["bandwidth_hz"].get<double>();
    if (bw > 0.0)
      return bw;
  }
  return std::nullopt;
}

json PM::describe() const {
  return {
      {"ports",
       {{"opt_in", "optical"},
        {"elec_in", "electrical"},
        {"opt_out", "optical"}}},
      {"spec",
       {{"params",
         {{"Vpi", {{"type", "float"}, {"default", 3.5}, {"unit", "V"}}},
          {"phi_bias", {{"type", "float"}, {"default", 0.0}, {"unit", "rad"}}},
          {"bandwidth_hz",
           {{"type", "float"}, {"default", 0.0}, {"unit", "Hz"}}}}},
        {"nonideal",
         {{"enable", {{"type", "bool"}, {"default", false}}},
          {"loss_db", {{"type", "float"}, {"default", 0.0}, {"unit", "dB"}}},
          {"vpi_error_pct",
           {{"type", "float"}, {"default", 0.0}, {"unit", "%"}}},
          {"drive_noise_rms",
           {{"type", "float"}, {"default", 0.0}, {"unit", "V"}}},
          {"bias_error_rad",
           {{"type", "float"}, {"default", 0.0}, {"unit", "rad"}}}}}}}};
}

REGISTER_BLOCK(PM, "PM");

} // namespace photonflow

File: backend/src/photonflow/blocks/optical/pm.py
"""Phase modulator block."""

from __future__ import annotations

import math
from typing import Dict

import torch

from photonflow.blocks.base import BaseBlock, register_block
from photonflow.core.filters import apply_lowpass
from photonflow.core.signal import Signal
from photonflow.core.sim import SimContext


@register_block("PM")
class PM(BaseBlock):
    """Electro-optic phase modulator. Applies phase modulation to an optical signal based on an electrical drive voltage. Key parameter: Vπ (half-wave voltage)."""

    PORTS = {"opt_in": "optical", "elec_in": "electrical", "opt_out": "optical"}
    SPEC = {
        "params": {
            "Vpi": {"type": "float", "default": 3.5, "unit": "V"},
            "phi_bias": {"type": "float", "default": 0.0, "unit": "rad"},
            "bandwidth_hz": {"type": "float", "default": 0.0, "unit": "Hz"},
            "bandwidth_kind": {
                "type": "enum",
                "default": "rect",
                "options": ["rect", "rc"],
            },
        },
        "nonideal": {
            "enable": {"type": "bool", "default": False},
            "loss_db": {"type": "float", "default": 0.0, "unit": "dB"},
            "vpi_error_pct": {"type": "float", "default": 0.0, "unit": "%"},
            "drive_noise_rms": {"type": "float", "default": 0.0, "unit": "V"},
            "bias_error_rad": {"type": "float", "default": 0.0, "unit": "rad"},
        },
    }

    def estimate_fmax(self) -> float | None:
        bw = self.params.get("bandwidth_hz")
        if bw is not None and float(bw) > 0.0:
            return float(bw)
        return None

    def process(self, inputs: Dict[str, Signal], ctx: SimContext) -> Dict[str, Signal]:
        opt_in = inputs["opt_in"]
        elec_in = inputs.get("elec_in")

        v = elec_in.data if elec_in is not None else torch.zeros_like(opt_in.data.real)
        v = torch.real(v)

        vpi = float(self.params.get("Vpi", 3.5))
        phi_bias = float(self.params.get("phi_bias", 0.0))
        bw_kind = self.params.get("bandwidth_kind", "rect")

        nonideal = self.nonideal if self.nonideal.get("enable", False) else {}
        vpi_error = float(nonideal.get("vpi_error", nonideal.get("vpi_error_pct", 0.0)))
        vpi = vpi * (1.0 + vpi_error / 100.0)
        bias_error = float(nonideal.get("bias_error_rad", 0.0))
        drive_noise_rms = float(nonideal.get("drive_noise_rms", 0.0))
        if drive_noise_rms > 0.0:
            v = v + drive_noise_rms * torch.randn_like(v)
        bandwidth_raw = self.params.get("bandwidth_hz")
        if bandwidth_raw is not None:
            bandwidth = float(bandwidth_raw)
            if 0.0 < bandwidth < ctx.fs / 2.0:
                v = apply_lowpass(v, ctx.fs, bandwidth, kind=bw_kind)

        phi = (phi_bias + bias_error) + math.pi * v / vpi
        phase = torch.exp(1j * phi)

        data = opt_in.data * phase
        loss_db = nonideal.get("loss_db")
        if loss_db is not None:
            data = data * (10 ** (-float(loss_db) / 20.0))

        signal = Signal(
            data=data,
            fs=opt_in.fs,
            t0=opt_in.t0,
            center_freq=opt_in.center_freq,
            pol_mode=opt_in.pol_mode,
            meta=dict(opt_in.meta),
        )
        return {"opt_out": signal}
